cpp/include/raft/common/nvtx.hpp

```diff
@@ -0,0 +1,119 @@
+
+namespace raft {
+namespace common {
+
+/**
+ * @brief Push a named nvtx range
+ * @param format range name format (accepts printf-style arguments)
+ * @param args the arguments for the printf-style formatting
+ */
+template <typename... Args>
+void PUSH_RANGE(const char* format, Args... args)
+{
+  detail::pushRange(format, args...);
+}
+
+/**
+ * @brief Synchronize CUDA stream and push a named nvtx range
+ * @param format range name format (accepts printf-style arguments)
+ * @param args the arguments for the printf-style formatting
+ * @param stream stream to synchronize
+ */
+template <typename... Args>
+void PUSH_RANGE(rmm::cuda_stream_view stream, const char* format, Args... args)
+{
+  detail::pushRange(stream, format, args...);
+}
+
+/** Pop the latest range */
+void POP_RANGE() { detail::popRange(); }
+
+/**
+ * @brief Synchronize CUDA stream and pop the latest nvtx range
+ * @param stream stream to synchronize
+ */
+void POP_RANGE(rmm::cuda_stream_view stream) { detail::popRange(stream); }
+
+/** Push a named nvtx range that would be popped at the end of the object lifetime. */
+class AUTO_RANGE {
+ private:
+  std::optional<rmm::cuda_stream_view> streamMaybe;
+
+  /* This object is not meant to be touched. */
+  AUTO_RANGE(const AUTO_RANGE&) = delete;
+  AUTO_RANGE(AUTO_RANGE&&)      = delete;
+  AUTO_RANGE& operator=(const AUTO_RANGE&) = delete;
+  AUTO_RANGE& operator=(AUTO_RANGE&&) = delete;
+
+ public:
+  /**
+   * Synchronize CUDA stream and push a named nvtx range
+   * At the end of the object lifetime, synchronize again and pop the range.
+   *
+   * @param stream stream to synchronize
+   * @param format range name format (accepts printf-style arguments)
+   * @param args the arguments for the printf-style formatting
+   */
+  template <typename... Args>
+  AUTO_RANGE(rmm::cuda_stream_view stream, const char* format, Args... args)
+    : streamMaybe(std::make_optional(stream))
+  {
+    PUSH_RANGE(stream, format, args...);
+  }
+
+  /**
+   * Push a named nvtx range.
+   * At the end of the object lifetime, pop the range back.
+   *
+   * @param format range name format (accepts printf-style arguments)
+   * @param args the arguments for the printf-style formatting
+   */
+  template <typename... Args>
+  AUTO_RANGE(const char* format, Args... args) : streamMaybe(std::nullopt)
+  {
+    PUSH_RANGE(format, args...);
+  }
+
+  ~AUTO_RANGE()
+  {
+    if (streamMaybe.has_value())
+      POP_RANGE(streamMaybe.value());
+    else
+      POP_RANGE();
+  }
+};
+
+/*!
+  \def RAFT_USING_RANGE(...)
+  When NVTX is enabled, push a named nvtx range and pop it at the end of the enclosing code block.
+
+  This macro initializes a dummy AUTO_RANGE variable on the stack,
+  which pushes the range in its constructor and pops it in the destructor.
+*/
+#ifdef NVTX_ENABLED
+#define RAFT_USING_RANGE(...) raft::common::AUTO_RANGE _AUTO_RANGE_##__LINE__(__VA_ARGS__)
+#else
+#define RAFT_USING_RANGE(...) (void)0
+#endif
```

Comment:
Tagging @wphicks , @cjnolet to determine the faith of the macro (continuing discussion https://github.com/rapidsai/cuml/pull/4436).

A quick recap; I added the macro in the first place because:
  1. It looks neater *to me* to use `RAFT_USING_RANGE("nvtx label")` instead of `AUTO_RANGE someName("nvtx label")`
  2. It generates a unique name, so I'm sure the compiler wouldn't generate any warnings about name shadowing in nested blocks.
  3. With NVTX disabled, I'm sure it compiles to absolutely nothing and I'm sure the compiler won't give an unused variable warning
 
Which isn't compelling to @wphicks, because we should avoid adding new macros when possible, and the compiler is smart enough to not do things mentioned in (3).